civihr_employee_portal/src/Security/PublicFirewall.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+namespace Drupal\civihr_employee_portal\Security;
+
+/**
+ * This class is responsible for checking if anonymous users can access
+ * certain routes
+ */
+class PublicFirewall {
+
+  /**
+   * Anonymous users can access these paths
+   *
+   * @var array
+   */
+  public static $publicRoutes = [
+    '/^welcome-page$/',
+    '/^sites\/default\/files\/logo.jpg$/', // if logo is missing
+    '/^request_new_account\/ajax$/', // from login page
+    '/^user((?!\/register).)*$/', // user* except user/register
+    '/^yoti-connect*/' // yoti login plugin
+  ];
+
+  /**
+   * @param \stdClass $user the user to check access for
+   * @param string $route the route to check
+   *
+   * @return bool
+   */
+  public function canAccess($user, $route) {
+    $isAnonymous = is_null($user) || $user->uid == 0;
+
+    if (!$isAnonymous) {
+      return TRUE;
+    }
+
+    return $this->allowAnonymousAccess($route);
+  }
+
+  /**
+   * Check if route matches against all public routes
+   *
+   * @param string $route
+   * @return bool
+   */
+  private function allowAnonymousAccess($route) {
+    foreach (self::$publicRoutes as $pattern) {
+      if (preg_match($pattern, $route)) {
```

Comment:
I did a quick benchmark, for 1,000,000 iterations the time was 4.1 seconds for all regex vs 2.9 for part-regex so I think that's worth changing, especially if this list grows in the future.

For our specific use case it would be safe to say that `/` means a regex since the path is coming from request_path() which trims leading and trailing slashes but I also think that using `/` to denote a regex when URLs are involved might be confusing.

I'm a bit wary of this becoming too complex if we introduce our own symbols  How about:

```
$publicRoutes = [
  'welcome-page',
  'sites/default/files/logo.jpg', // if logo is missing
  'request_new_account/ajax', // from login page
  '@^user((?!\/register).)*$@', // user* except user/register
  '@^yoti-connect*@' // yoti login plugin
];
```

Where we would use `@` as the regex delimiter. I think most developers familiar with regex would understand without any further explanation.